### PR TITLE
*: port table and refactor properties

### PR DESCRIPTION
**Pull request by BusyJay**
This PR ports most table related types and functions.

It also adds a macros `simple_access` to generate getters/setters. It can guarantee no mistakes when repeating implementing new getters/setters and make code easier to view (except the macro itself unfortunately).

A trait `PathToSlice` is also added to make convert `Path` to `rocksdb::Slice` easier on Unix.

And naming are unified now that `as_ptr` takes `&` and outputs `*const`, `as_mut_ptr` takes `&mut` and outputs `*mut`, `get_ptr` takes `&` and outputs `*mut`.

There are still some places that gets polish.

A full version of changes can be found at dev branch.

### CODE REVIEW

File: tirocks/src/lib.rs [hunk 1]
```diff
@@ -1,19 +1,32 @@
...
-pub mod table_properties;
+pub mod table;
+pub mod table_filter;
 mod util;
 
 pub use error::{Code, Result, Severity, Status, SubCode};
+use tirocks_sys::rocksdb_DB;
+
+// TODO: define later.
+pub struct RawDb;
```
Comment (tabokie) [inline tirocks/src/lib.rs:26]: Why is it called raw? Unless you plan to have a more high-level DB in this crate?

File: tirocks/src/option/read.rs [hunk 2]
```diff
@@ -253,4 +303,20 @@ impl ReadOptions {
...
+    #[inline]
+    pub(crate) fn get_ptr(&self) -> *const rocksdb_titandb_TitanReadOptions {
+        &self.raw
+    }
 }
+
+impl Drop for ReadOptions {
+    fn drop(&mut self) {
+        unsafe {
+            tirocks_sys::ctitandb_readoptions_destroy(&mut self.raw);
```
Comment (tabokie) [inline tirocks/src/option/read.rs]: I think it should be renamed to "inplace_destroy" to avoid confusion.

File: tirocks/src/properties.rs [hunk 3]
```diff
@@ -0,0 +1,556 @@
...
+}
+
+pub struct PropertyMap {
+    ptr: *mut crocksdb_map_property_t,
+}
+
+impl Drop for PropertyMap {
+    #[inline]
+    fn drop(&mut self) {
+        unsafe { tirocks_sys::crocksdb_destroy_map_property(self.ptr) }
```
Comment (tabokie) [inline tirocks/src/properties.rs]: Naming is inconsistent:
```suggestion
        unsafe { tirocks_sys::crocksdb_map_property_destroy(self.ptr) }
```

File: tirocks/src/cache.rs [hunk 4]
```diff
@@ -0,0 +1,193 @@
+// Copyright 2022 TiKV Project Authors. Licensed under Apache-2.0.
+
+use tirocks_sys::{crocksdb_cache_t, rocksdb_LRUCacheOptions};
+
+use crate::util::simple_access;
+
+#[cfg(feature = "jemalloc")]
+#[derive(Debug)]
+#[repr(transparent)]
+pub struct JemallocAllocatorOptions(tirocks_sys::rocksdb_JemallocAllocatorOptions);
```
Comment (tabokie) [inline tirocks/src/cache.rs:10]: Missing Drop implementation.

File: tirocks/src/util.rs [hunk 5]
```diff
@@ -23,6 +28,93 @@ macro_rules! utf8_name {
...
+/// impl A {
+///     simple_access! {
+///         prefix
+///         /// Docs
+///         name: Type
+///     }
+/// }
+/// ```
+/// `simple_access` will generate setter `set_name` with the default implementation call
+/// to `prefix_set_name(self.as_mut_ptr(), val)`.
```
Comment (tabokie) [inline tirocks/src/util.rs]: Adding some more examples for advanced use case? e.g. (<) [.get_ptr()]

File: tirocks/src/lib.rs (hunk 1 above)
Comment (BusyJay) [reply tirocks/src/lib.rs:26]: Yes. `RawDb` is the struct that only contains pointer. `Db` will be the struct that also contains `cf`, `env` and `comparater`. `Db` is the only struct provides lifetime safety.

File: tirocks/src/cache.rs (hunk 4 above)
Comment (BusyJay) [reply tirocks/src/cache.rs:10]: It's a struct that doesn't need drop.

File: tirocks/src/table/plain.rs [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+#[repr(transparent)]
+pub struct PlainTableOptions(rocksdb_PlainTableOptions);
+
+impl PlainTableOptions {
+    /// plain table has optimization for fix-sized keys, which can
+    /// be specified via user_key_len.  Alternatively, you can pass
+    /// `kPlainTableVariableLength` if your keys have variable
+    /// lengths.
+    #[inline]
+    pub fn set_user_key_len(&mut self, len: u32) -> &mut Self {
```
Comment (Connor1996) [inline tirocks/src/table/plain.rs:16]: why not written in `simple_access` manner

Comment (BusyJay) [reply tirocks/src/table/plain.rs:16]: Because it doesn't need C function.

Comment (BusyJay) [reply tirocks/src/table/plain.rs:16]: Added the use condition to the comment of `simple_access`.

**Review by tabokie — APPROVED** [review]

**Review by Connor1996 — APPROVED** [review]
LGTM

**Pull request merged**